scripts/providers/EbookFoundationFPB/process.py

```diff
@@ -0,0 +1,116 @@
+import time
+
+import requests
+
+from openlibrary.plugins.upstream.utils import get_marc21_language
+
+
+def fetch_data_from_ebookfoundation(url, max_retries=10, delay=5):
+    """Fetch JSON from a URL, retrying until a successful response is received."""
+    attempt = 0
+
+    while True:
+        try:
+            response = requests.get(url, timeout=10)
+            response.raise_for_status()  # Raise an error if request fails
+            return response.json()  # Successfully fetched JSON
+
+        except requests.exceptions.RequestException as e:
+            attempt += 1
+            print(
+                f"Attempt {attempt}: Failed to fetch JSON ({e}). Retrying in {delay} seconds..."
+            )
+            time.sleep(delay)  # Wait before retrying
+
+            if attempt >= max_retries:
+                print("Max retries reached. Exiting...")
+                return None  # Return None if max retries are exceeded
+
+
+def flatten_books(data):
+    flat_list = []
+
+    def traverse(sections, topics, language_code):
+        for section in sections:
+            new_topics = topics + [section["section"]]
+
+            # Process entries at the current section level
+            for book in section.get("entries", []):
+                if "url" not in book:
+                    continue
+
+                authors = []
+                raw_authors = (
+                    book["author"].split(",") if ("author" in book) else ["????"]
+                )
+
+                for author in raw_authors:
+                    if not author:
+                        continue
+
+                    authors.append({"name": author.strip()})
+
+                notes = "".join(book.get("notes", []))
+                format = ""
+
+                if "pdf" in notes.lower():
+                    format = "pdf"
+                elif "epub" in notes.lower():
+                    format = "epub"
+                else:
+                    format = "web"
+
+                flat_list.append(
+                    {
+                        "title": book.get("title", "????"),
+                        "authors": authors,
+                        "source_records": ["????"],
```

Comment:
```suggestion
                        "source_records": ["EbookFoundation"],
```
This might be an improvement on `"????"`.

Reply:
I have changed this to this new suggestion.